x-pack/plugins/secret_service/server/service/secret_service.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+// @ts-ignores
+import nodeCrypto from '@elastic/node-crypto';
+import crypto from 'crypto';
+import { SavedObjectsClient } from 'src/legacy/server/saved_objects';
+// @ts-ignore
+import { AuditLogger } from '../../../server/lib/audit_logger';
+
+type SSEvents = 'secret_object_created' | 'secret_object_accessed' | 'secret_object_decrypt_failed';
+type SecretId = Promise<string>;
+
+export class SecretService {
+  public readonly hideAttribute: (deets: any) => SecretId;
+  public readonly unhideAttribute: (deets: any) => any;
+  public readonly validateKey: () => Promise<boolean>;
+  private readonly type: string;
+  private readonly auditor?: AuditLogger;
+  private readonly hideAttributeWithId: (id: string, toEncrypt: any) => SecretId;
+
+  constructor(
+    savedObjectsClient: SavedObjectsClient,
+    type: string,
+    key?: Buffer,
+    auditor?: AuditLogger
+  ) {
+    this.type = type;
+    this.auditor = auditor;
+    key = key || crypto.randomBytes(128);
+    const crypt = nodeCrypto({ encryptionKey: key.toString('hex') });
+
+    const logEvent = (event: SSEvents, message?: string) => {
+      if (this.auditor) {
+        this.auditor.log(event, message);
+      }
+    };
+
+    this.hideAttribute = async (toEncrypt: any) => {
+      return this.hideAttributeWithId(crypto.randomBytes(32).toString('base64'), toEncrypt);
```

Comment:
Good call on using our own specified ID; however, to remain consistent with ES and ensure we aren't introducing bias with how we're converting the number, let's use a v4 uuid here.

Reply:
Interesting, looking at the saved object repository they are using `v1` so we are already not consistent with ES not sure what consistency buys us here. But I don't see a problem using v4 from the npm package it does use `crypto.randomBytes(16)` versus `32` both have plenty of range.